### ISSUE TITLE
Basic commands

### ISSUE THREAD

**Issue by krystophny** [itpplasma/dat#1]
- Commit: Create an entry with metadata
- Push: 
  - Copy the finished run from temporary to main data directory
  - Update metadata record
- Pull:
  - Copy data from main directory to temp directory
- Archive: 
  - Move run from main data directory to archive
  - Update metadata record

No delete, no overwrite, just write previously non-existing files for now. If you delete them, they can be updated.

Main data directory is also accessible via shared storage to collaborators (e.g. Nextcloud).

Inside ITPcp, copy between `/temp` and `/proj/plasma` is easiest. Outside, Nextcloud is easiest.

**Comment by krystophny** [comment]
*Automated triage pass (Claude), grounded in the repository at `fe6c66f`. Verify before acting.*

Still relevant, and still entirely unimplemented. Of the four commands specified here, `commit` exists in partial, partly-stubbed form; `push`, `pull` and `archive` do not exist at all and crash with a bare `KeyError`. `fe6c66f` is the tip of `main` and the last commit in the repository (2024-02-15) — no refactor has superseded this issue.

## Purpose

ITPcp simulation runs are produced on scratch (`/temp`), have to end up on group storage (`/proj/plasma`), and have to be reachable by external collaborators through Nextcloud. Today that is done by hand with `cp`/`rsync`, so provenance — which code, which inputs, which run produced which directory — lives in people's heads and in ad-hoc `README` files. This issue defines the four verbs that make a run's lifecycle explicit and auditable: `pull` inputs, `commit` metadata at launch, `push` results, `archive` when cold. The append-only constraint in the issue body ("no delete, no overwrite, just write previously non-existing files") is the safety property that makes this usable on shared group storage without a review gate: a mistaken `push` can never destroy a colleague's results.

## Current state

The whole package is four files, 145 lines. Relevant facts:

`src/dat/dat.py:9-16` — dispatch is `globals()` lookup:

```python
def main():
    print("This is Dat.")
    command = get_function_by_name(sys.argv[1])
    command(sys.argv[2:])


def get_function_by_name(name):
    return globals()[name]
```

`commit` at `src/dat/dat.py:19-32` is the only command function at module scope, so `push`, `pull` and `archive` raise `KeyError`. The lookup also exposes every module global as a "command" (`os`, `sys`, `entry`, `local_storage`, `create_entry`, `has_title`, `JSONStorage`).

`src/dat/dat.py:6-7` runs at import time and hard-codes the metadata path to the CWD:

```python
entry = create_entry()
local_storage = JSONStorage("metadata.json")
```

`src/dat/dat.py:27` is missing the `f` prefix and prints literal braces. `src/dat/dat.py:30` is a placeholder print where the eLabFTW write belongs. `commit(args)` at `src/dat/dat.py:19` accepts `args` and ignores it, so the `-m <message>` flag from #2 is unimplemented.

`src/dat/journal.py:28-84` holds the intended architecture: a `Storage` strategy base with `ElabStorage` (implemented, **never instantiated anywhere**), `HDF5Storage` (both methods `raise NotImplementedError`), and `JSONStorage` (the only working backend). `Storage.load(self, entry)` and `JSONStorage.load(self)` have inconsistent signatures.

There is **no** config mechanism for the DATA/RUN/ARCHIVE locations that this issue is entirely about — no config file, no env var, nothing. `src/dat/draft.py` is dead scratch (references `argv`, `json`, `manager` with no imports) and should not be built on. There are no tests, no CI, no `.github/`.

## Reproduction / evidence

```bash
git clone https://github.com/itpplasma/dat && cd dat
python -m venv .venv && . .venv/bin/activate
pip install -e .

mkdir -p /tmp/run && cd /tmp/run
dat push
```

Observed (verified against `fe6c66f`):

```
This is Dat.
Traceback (most recent call last):
  ...
  File ".../src/dat/dat.py", line 11, in main
    command = get_function_by_name(sys.argv[1])
  File ".../src/dat/dat.py", line 16, in get_function_by_name
    return globals()[name]
KeyError: 'push'
```

Identical for `dat pull <path>` and `dat archive`. `dat` with no argument raises `IndexError` at `dat.py:11`. `dat os` gets past the lookup and dies with `TypeError: 'module' object is not callable`, demonstrating the namespace leak.

`dat commit` does run — it writes `{"title": "My run"}` to `./metadata.json` and prints the unformatted line:

```
Title: Committing changes to {local_storage.filename}...
Committing changes to eLabFTW...
Changes committed.
```

Expected after this issue: all four verbs dispatch, `dat push` refuses to overwrite an existing destination file with a non-zero exit code, and `metadata.json` carries a `status` field that advances `running → finished → archived`.

## Behaviour (BDD)

```gherkin
Scenario: Unknown command fails cleanly instead of KeyError
  Given a shell in any directory
  When I run "dat frobnicate"
  Then stderr contains "unknown command: frobnicate"
  And stderr lists "commit, pull, push, archive"
  And the exit code is 2
  And no traceback is printed

Scenario: Module globals are not commands
  Given a shell in any directory
  When I run "dat os"
  Then stderr contains "unknown command: os"
  And the exit code is 2

Scenario: Pull copies inputs from DATA into the run directory
  Given DAT_DATA="/proj/plasma" containing "/proj/plasma/demo/input.in"
  And an empty current working directory "/temp/albert/run01"
  When I run "dat pull demo"
  Then "/temp/albert/run01/input.in" exists with byte-identical content to "/proj/plasma/demo/input.in"
  And "/proj/plasma/demo/input.in" is unmodified (same mtime and size)
  And "./metadata.json" has "source" == "demo"

Scenario: Commit records a started run
  Given a current working directory with no "metadata.json"
  When I run "dat commit -m 'ITER TC24 scan, n=2'"
  Then "./metadata.json" exists and parses as JSON
  And it contains "title" == "ITER TC24 scan, n=2"
  And it contains "status" == "running"
  And it contains "created" matching an ISO-8601 timestamp
  And no interactive prompt is shown

Scenario: Push is append-only and refuses to overwrite
  Given "./metadata.json" with "status" == "running" and "target" == "demo/run01"
  And "/proj/plasma/demo/run01/result.h5" already exists
  When I run "dat push"
  Then the exit code is 1
  And stderr names "/proj/plasma/demo/run01/result.h5" as an existing destination file
  And "/proj/plasma/demo/run01/result.h5" is unmodified
  And "./metadata.json" still has "status" == "running"

Scenario: Archive moves the run and updates the record
  Given "/proj/plasma/demo/run01" exists with "status" == "finished" in its metadata.json
  And DAT_ARCHIVE="/proj/plasma/archive"
  When I run "dat archive demo/run01"
  Then "/proj/plasma/archive/demo/run01/result.h5" exists
  And "/proj/plasma/demo/run01" no longer exists
  And "/proj/plasma/archive/demo/run01/metadata.json" has "status" == "archived"
```

## Implementation plan

Do the prerequisites (1-3) first; they are small and every command depends on them.

1. **`src/dat/config.py` (new).** Add `load_config()` returning a `Config` dataclass with fields `data` (DATA root), `archive` (archive root), `remote` (optional rclone remote name, `None` for plain filesystem). Resolution order: environment variables `DAT_DATA`, `DAT_ARCHIVE`, `DAT_REMOTE`; then `~/.config/dat/config.json`; then error with an actionable message. Document `/proj/plasma` and `/proj/plasma/archive` as the ITPcp deployment example, not as portable defaults. Raise `ConfigError` (defined here) if `data` is unset.

2. **`src/dat/dat.py` — replace `get_function_by_name`.** Delete `get_function_by_name` (`dat.py:15-16`) and the import-time globals (`dat.py:6-7`). Introduce an explicit `COMMANDS = {"commit": commit, "pull": pull, "push": push, "archive": archive}` dict and dispatch through `COMMANDS.get(name)`. On miss, print `unknown command: <name>` plus the sorted key list to stderr and `sys.exit(2)`. On bare `dat` (no `sys.argv[1]`), print the same usage and `sys.exit(2)`. Move `entry`/`local_storage` construction inside each command so the metadata path is resolved per invocation, not at import.

3. **`src/dat/dat.py:27` — fix the missing `f` prefix.** `print(f"Committing changes to {local_storage.filename}...")`.

4. **`src/dat/journal.py` — extend the entry schema.** Change `create_entry(title="Untitled")` (`journal.py:10-13`) to also set `"status": "running"`, `"created": datetime.now(timezone.utc).isoformat()`, `"source": None`, `"target": None`, `"host": socket.gethostname()`, `"user": getpass.getuser()`. Add `set_status(entry, status)` validating against `{"running", "finished", "archived"}`. Fix the signature mismatch by making the base `Storage.load(self)` take no `entry` argument, matching `JSONStorage.load` (`journal.py:75`); update `ElabStorage.load` and `HDF5Storage.load` accordingly.

5. **`src/dat/transport.py` (new).** Mirror the `Storage` strategy idiom from `journal.py`. Define `Transport` with `copy(src, dst)` and `move(src, dst)`; then:
   - `LocalTransport` — walks `src`, and for every file computes the destination path; if it exists, collect it into a list and **raise `OverwriteError` listing all collisions before copying anything** (check-then-act, no partial writes). Use `shutil.copy2` for the actual copy.
   - `RcloneTransport` — construct destinations explicitly as `<remote>:<root>/<relative-path>`, preflight the destination listing, and fail on any collision before invoking a transfer. `--immutable` may be used as a second guard, but it is not a substitute for the collision preflight; do not use `--error-on-no-transfer`, which also rejects an idempotent copy. Surface a non-zero rclone exit as a transfer error. Do not use `rclone sync` anywhere — it deletes on the destination, which this issue forbids.
   Select the implementation from `Config.remote`. For the first PR, keep the append-only collision tests on `LocalTransport`; remote metadata updates need an explicit staged-copy protocol before `RcloneTransport.move` is declared complete.

6. **`src/dat/dat.py` — add `pull(args)`.** Signature `dat pull <path>` where `<path>` is DATA-relative. Copy `<DATA>/<path>` into the CWD via `Transport.copy`. Load-or-create `metadata.json`, set `entry["source"] = path`, save. Never touch DATA.

7. **`src/dat/dat.py` — add `-m` handling to `commit(args)`.** Parse with `argparse` in a helper `parse_commit_args(args)`; `-m/--message` sets the title non-interactively. Keep the `input("Title: ")` fallback only when `-m` is absent **and** `sys.stdin.isatty()`; otherwise exit 2 with "no title: pass -m". Set `status="running"`. Leave the eLabFTW call out of scope here — `dat.py:30` stays a placeholder and is tracked by #2.

8. **`src/dat/dat.py` — add `push(args)`.** Read `metadata.json` from the CWD (error if missing: "nothing to push, run dat commit first"). Determine the destination as `<DATA>/<entry["target"]>`, defaulting `target` to `<source>/<basename(cwd)>` when unset. Set `status="finished"` and `"pushed"` timestamp **and save `metadata.json` locally first**, then `Transport.copy(cwd, dest)` so the pushed copy carries the finished status. If `copy` raises `OverwriteError`, revert the local status to `running`, print the collision list, exit 1.

9. **`src/dat/dat.py` — add `archive(args)`.** Signature `dat archive <path>`, DATA-relative. Refuse unless the record at `<DATA>/<path>/metadata.json` has `status == "finished"` (exit 1 otherwise). For the local transport, move to `<ARCHIVE>/<path>` only after a complete collision preflight, then rewrite the moved `metadata.json` with `status="archived"` and an `"archived"` timestamp. For rclone, stage the updated metadata and transfer the complete tree before removing the DATA copy; do not claim the local rewrite-after-move sequence is safe for a remote URI until that protocol is tested. Same collision guard: refuse if `<ARCHIVE>/<path>` already exists.

10. **`tests/` (new directory) — establish the test idiom, there is none today.** Add `tests/test_dispatch.py` (unknown command exits 2; `dat os` exits 2; bare `dat` exits 2), `tests/test_transport.py` (`LocalTransport.copy` into a `tmp_path` with a pre-existing file raises `OverwriteError` and leaves the destination byte-identical), `tests/test_commands.py` (full `pull → commit → push → archive` round trip against two `tmp_path` roots wired via monkeypatched `DAT_DATA`/`DAT_ARCHIVE`, asserting the `status` transitions). Use `pytest` with `tmp_path` and `monkeypatch` only — do not require a live eLabFTW server or `rclone` in the test suite; mark any rclone test `@pytest.mark.skipif(shutil.which("rclone") is None, ...)`.

11. **`pyproject.toml` — wire the tests.** Add `[project.optional-dependencies] dev = ["pytest>=7"]` and `[tool.pytest.ini_options] testpaths = ["tests"]`. The `[project.scripts] dat = "dat.dat:main"` entry point is already correct and needs no change.

12. **`.github/workflows/ci.yml` (new).** There is no CI at all. Minimal: `ubuntu-latest`, Python 3.11, `pip install -e .[dev]`, `pytest`. Cheap, and without it the append-only guarantee has nothing defending it.

13. **`README.md` — replace the two-line stub** with the four commands, the `DAT_DATA`/`DAT_ARCHIVE`/`DAT_REMOTE` variables, and the two ITPcp deployment recipes from this issue (filesystem `/temp` ↔ `/proj/plasma`; rclone remote for Nextcloud).

No sibling repo in the org needs to change. `dat` has no submodules, no `FetchContent`, no fpm or path dependencies — the only external couplings are the PyPI package `elabapy` and, per #2, the `rclone` binary.

## Acceptance criteria

- [ ] `pip install -e .[dev] && pytest` passes from a clean checkout.
- [ ] `dat`, `dat frobnicate` and `dat os` each exit with code 2 and print a usage line naming `commit, pull, push, archive`. No `KeyError`, no `IndexError`, no traceback.
- [ ] `dat commit -m "msg"` is fully non-interactive: it completes with stdin closed (`dat commit -m x < /dev/null`) and writes `status: "running"`.
- [ ] `dat push` into a destination holding an identically-named file exits 1, prints the colliding path, and leaves that file byte-identical (verified by `sha256sum` before and after) — asserted in `tests/test_transport.py`.
- [ ] A `pull → commit → push → archive` round trip over two `tmp_path` roots leaves `status == "archived"` in the archived `metadata.json` and no directory at the DATA location — asserted in `tests/test_commands.py`.
- [ ] `rg -n "globals\(\)" src/` returns nothing.
- [ ] `rg -n 'print\("Committing changes to \{' src/` returns nothing (the f-string bug is gone).
- [ ] No `rclone sync` anywhere: `rg -n "sync" src/` returns no rclone invocation.
- [ ] CI is green on `main`.

## References

- Davison, A. P. (2012), *Automated capture of experiment context for easier reproducibility in computational research*, Computing in Science & Engineering 14(4), 48-56, doi:10.1109/MCSE.2012.41 — Sumatra, the closest prior art: it records each simulation run's code version, parameters and outputs as a "record" and offers exactly this `smt run`/`smt list` verb style. Worth reading before finalising the `metadata.json` schema in step 4.
- Halchenko, Y. O. et al. (2021), *DataLad: distributed system for joint management of code, data, and their relationship*, Journal of Open Source Software 6(63), 3262, doi:10.21105/joss.03262 — solves the same DATA↔RUN problem on top of git-annex. Its `datalad get`/`datalad push` semantics are the reference model for the pull/push pair here; note that DataLad deliberately makes annexed files read-only in the working tree, which is the same append-only property this issue asks for, enforced by the filesystem rather than by a check.
- Carpi, N. et al. (2017), *eLabFTW: An open source laboratory notebook for research labs*, Journal of Open Source Software 2(12), 146, doi:10.21105/joss.00146 — the notebook backend `src/dat/journal.py:1-4` targets.
- Wilkinson, M. D. et al. (2016), *The FAIR Guiding Principles for scientific data management and stewardship*, Scientific Data 3, 160018, doi:10.1038/sdata.2016.18 — the F (findable, via the eLabFTW record) and A (accessible, via the Nextcloud mirror) motivation for the shared-storage requirement in this issue.
- `rclone` (https://rclone.org) — the transfer backend proposed in #2. The relevant flags for step 5 are `copy` (never `sync`) and `--immutable`.

## Scope and risk

Steps 1-3 plus 10-12 are one small PR (dispatch, config, first tests, CI) and unblock everything else; steps 4-9 are one command each and are cleanly separable. **Recommend splitting** into five atomic issues: *dispatch + config + CI scaffold*, then *pull*, *commit `-m`*, *push*, *archive* — this issue then becomes the tracker. Main risk is `push` and `archive` operating on real group storage: the collision check must be complete-before-copy (step 5) and `archive`'s move must be the last operation, or a half-finished transfer leaves a run in neither location.

Related: itpplasma/dat#2 is the user story that this issue's verbs serve, and it owns the eLabFTW/`rclone`/`json-fortran` integration. The two are complementary, not duplicates — #1 defines command semantics, #2 defines the workflow and the database side. Keep the eLabFTW write out of this issue.

One thing to check before anyone invests in the `ElabStorage` path in `src/dat/journal.py:36-57`: `elabapy` (pinned in `pyproject.toml` and `requirements.txt`) is the older eLabFTW API v1 client, and upstream has since published `elabapi-python` for API v2. I could not verify the version running at elabftw.tugraz.at from this repository, so confirm the server's API version first — that decision belongs to #2, not here.
